feat(adr): SecretProvider for all EdgeX Services (#281)
edgexfoundry/edgex-docs: docs_src/design/adr/014-Secret-Provider-For-All.md

**Service Shared** secrets are those that all instances of a class of service ,such a Application Services, share. Think of Core Data as it own class of service. An example of shared secrets are the database credentials for the single database instance for Store and Forward data that all Application Services may need to access. Another example is the database credentials for each of instance the Core Data. It is shared, but only one instance of Core Data is currently ever run. Service Shared secrets are seeded by security-secretstore-setup using static configuration for known secrets for known services. Currently database credentials are the only shared secrets. In the future we may have Message Bus credentials as shared secrets, but these will be truly shared secrets for all services to securely connect to the Message Bus, not just shared between instances of a service.

Application Services currently have the ability to configure `SecretStores` for **Service Exclusive** and/or **Service Shared** secrets depending on their needs.

> #### Known and Unknown Services 

Author: > "static services and nonstatic services

Not a fan. Services are never static. They are staticky known via config, thus the Know vs Unkown.

  - Application Service (instance) are examples of these services. 

  - Service exclusive `SecretStore` can be created for these services by adding the services' unique name , i.e. appservice-http-export, to the `ADD_SECRETSTORE_TOKENS` environment variable for security-secretstore-setup

    ```
    ADD_SECRETSTORE_TOKENS: "appservice-http-export, appservice-mqtt-export"
    ```

    This creates an exclusive `SecretStore` for each service listed. The name provided for each service must be used in the service's `SecretStore` configuration and Docker volume mount. Typically the configuration is set via environment overrides or is already in an existing configuration profile (***http-export*** profile for app-service-configurable). 

    Example docker-compose file entries:

    ```yaml
    environment:
        ...
    	SecretStoreExclusive_Path: "/v1/secret/edgex/appservice-http-export/"
    	TokenFile: "/tmp/edgex/secrets/appservice-http-export/secrets-token.json"
    	
    volumes:
    	...
    	- /tmp/edgex/secrets/appservice-http-export:/tmp/edgex/secrets/appservice-http-export:ro,z
    ```

> #### Known and Unknown Secrets

Author: > "static Secrets and nonstatic Secrets"

As with services, secrets are not static. They are `known` statically via configuration.

}
```

###### Factory and bootstrap handler methods

```go
type SecretProviderImpl struct {
	SharedSecretClient    pkg.SecretClient
	ExclusiveSecretClient pkg.SecretClient
	secretsCache          map[string]map[string]string // secret's path, key, value
	configuration         *common.ConfigurationStruct
	cacheMuxtex           *sync.Mutex
	loggingClient         logger.LoggingClient
	//used to track when secrets have last been retrieved
	LastUpdated time.Time
}

func NewSecretProvider(
    loggingClient logger.LoggingClient, 
    configuration *common.ConfigurationStruct) *SecretProviderImpl {
	sp := &SecretProviderImpl{
		secretsCache:  make(map[string]map[string]string),
		cacheMuxtex:   &sync.Mutex{},
		configuration: configuration,
		loggingClient: loggingClient,
		LastUpdated:   time.Now(),
	}

	return sp
}
```

```go
type Secrets struct {
}

func NewSecrets() *Secrets {
	return &Secrets{}
}

func (_ *Secrets) BootstrapHandler(
	ctx context.Context,
	_ *sync.WaitGroup,
	startupTimer startup.Timer,
	dic *di.Container) bool {
    ...
    Creates NewNewSecretProvider, calls Initailizes() and adds it to the DIC
    ...
}
```

#### Secret Store for non-secure mode

> Both Bootstrap's and App SDK's implementation use the `DatabaseInfo` configuration for `GetDatabaseCredentials` API in non-secure mode. The App SDK only uses it, for backward compatibility,  if the database credentials are not found in the new `InsecureSecrets` configuration section. For Ireland it was panned to only use the new `InsecureSecrets` configuration section in non-secure mode.

Reviewer: You mean "For Ireland it is planned"? (not panned)

```yaml
ADD_SECRETSTORE_TOKENS: "<service-name1>,<service-name2>"
```

If we expanded this to add an optional list of known secret identifiers for each service, i.e. `Databases.application-service`, the exclusive store could also be seeded with a copy of known shared secrets. In this case the database credentials for the Application Services' shared database. The environment variable name will change to `ADD_SECRETSTORE` now that it is more than just tokens.

```yaml
ADD_SECRETSTORE: "app-service-xyz[Databases.application-service]"
```

Full specification for the environment variable's value is a comma separated list of service entries defined as:

```
<service-name>[optional list of known secret IDs sperated by ;]>
```

Example with one service specifying IDs for know secrets and one without known secrets 

```yaml
ADD_SECRETSTORE: "appservice-xyz[Databases.application-service; <other TBD id>], appservice-http-export"
```

> Once security-secretstore-setup generates the known application-service database credentials, it will save them in memory while processing additional adds of `SecretStores` so they can be used if other Application Services needs them seeded into their `SecretStores` . 

Reviewer: ## What actually happens today

1. security-secretstore-setup spawns a separate executable, security-file-token-provider.
1. the file-token-provider enumerates a configuration file and creates a policy and token for each service using its service key.  This token give access its the services' private area, e.g. secrets/edgex/edgex-core-data, as well as any custom paths carried over for legacy purposes, e.g. secret/edgex/coredata/mongodb
1. later on, we create database credentials by enumerating the configuration.toml and create a database secret for each database where a Service= key is specified.  For example: secret/edgex/coredata/mongodb  when Service=coredata.  There is also a copy of the secrets under secret/edgex/mongo or under secret/edgex/bootstrap-redis.  For mongodb, it uses the section name, such as application-service instead of the service tag, appservice. secret/edgex/mongo/application-service contains the same secret as secret/edgex/appservice/mongodb.  The reason for the duplication is that the database bootstrapper can enumerate the entire secret/edgex/mongo subtree when provisioning accounts.

## What will happen in the bootstrapping ADR

Database credentials will be bootstrapped as part of the install.  security-file-token-provider will run every boot to regenerate tokens.

## What this means for the above

security-file-token-provider doesn't really know anything about databases.  It just knows about tokens and policies.  Similarly, for legacy reasons the old code isn't even putting the database secrets in the right place to begin with -- where is inconsistent use of the service key in the framework.  (continue later)

Author: @bnevis-i,  I like where you are going. I think we can drop `duplicate` since any secret specified here is to be duplicated. Also since `/secret/edgex/` is always present, we could drop it also. This would make it nice and concise as:
 ```
app-service-xyz[appservice/redisdb]
```
If we also have credentials for message bus if would look something like:
```
app-service-xyz[appservice/redisdb; common/messagebus]
```
This would create the following for `edgex-app-service-xyz`
```
/secret/edgex/edgex-app-service-xyz/redisdb
/secret/edgex/edgex-app-service-xyz/messagebus
```
and assumes the following new policy add to `token-config-json` :
```
  "edgex-service-common": {
    "edgex_use_defaults": true,
    "custom_policy": {
      "path": {
        "secret/edgex/common/messagebus": {
          "capabilities": [
            "list",
            "read"
          ]
        },
      }
    }
  },
```

This all depends on the database or other credential bootstrapping have completed and stored the credentials before the environment variable is processed. Don't think it happening in that order now, but we could refactor to ensure it is. Thoughts?

Reviewer: This makes sense.

However, `token-config.json` is just for creating tokens.  `token-config.json` doesn't need to be altered as per above unless one is creating a `service-common` service that needs its own token.  All of the known secret seeding today is done with a root token that has access to the entire KV store.

Author: @bnevis-i , My thoughts with `service-common` is it is similar to `appservice`. There isn't a specific service called AppService, but we create the token and secret store for it, put the app service shared secrets in it and then use it to copy the shared secrets to actual App Services secret stores when they declare themselves via the environment variable.

`service-common` would be there for shared secrets that all edgex services need, such as the message bus credentials. Does this make sense?

Reviewer: I'm not opposed to the idea of `service-common`.  I am just suggesting that we might not need a separate token for it.  security-secretstore-setup is trusted to read/write secrets anywhere in the URL space, whether it is covered by a token or not.

### Implementation

#### Factory Method and Bootstrap Handler

The factory method and bootstrap handler will follow that currently in the Bootstrap implementation with some tweaks. Rather than putting the two split interfaces into the DIC, it will put just the single interface instance into the DIC. See details in the [Interfaces and factory methods](#interfaces-and-factory-methods) section above under **Existing Implementations**.

#### Caching of Secrets

Secrets will be cached as they are currently in the Application Service implementation

#### Insecure Secrets

Insecure Secrets will be handled as they are currently in the Application Service implementation. `DatabaseInfo` configuration will no longer be an option for storing the insecure database credentials. They will be stored in the `InsecureSecrets` configuration only.

```toml
[Writable.InsecureSecrets]
    [Writable.InsecureSecrets.DB]
    	path = "redisdb"
    	[Writable.InsecureSecrets.DB.Secrets]
            username = ""
            password = ""
```

##### Handling on-the-fly changes to `InsecureSecrets`

> All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap` as along with existing log level processing. This special processing will be taken from App SDK.

Reviewer 2: ```suggestion
All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap`  along with existing log level processing. This special processing will be taken from App SDK.
```

#### Mocks

> Proper mock of the `SecretProvider` interface will be created with `Mockery` to be used in unit tests. Current mock in App SDK is handle built.

Reviewer: What does this sentence mean: "Current mock in App SDK is handle built."

Author: It is hand written, rather then generated using `Mockery` so the API response can not be customized by the unit test as they can with mocks generated with `Mockery`. Will reword to `Current mock in App SDK is handle written rather then generated with Mockery`.

Refactoring the existing implementation in `go-mod-bootstrap` and have it reside there now seems to be the best choice.

### Consequences

- All service's will have `Writable.InsecureSecrets` section added to their configuration
- `InsecureSecrets` definition will be moved from App SDK to go-mod-bootstrap
- edgex-go services will be modified to use the single `SecretProvider` interface from the DIC in place of current usage of the `GetDatabaseCredentials` and `GetCertificateKeyPair` interfaces.
  - Calls to `GetDatabaseCredentials` and `GetCertificateKeyPair` will be replaced with calls to `GetSecrets` API and appropriate processing of the returned secrets will be added. 
- App SDK will be modified to use `GetSecrets` API in place of the `GetDatabaseCredentials` API
- App SDK will be modified to use the new `SecretProvider` bootstrap handler

> - app-service-configurable's configuration profiles will be updated to remove the `SecretStoreExclusive` configuration and just us the existing `SecretStore` configuration

Reviewer 2: ```suggestion
- app-service-configurable's configuration profiles will be updated to remove the `SecretStoreExclusive` configuration and just use the existing `SecretStore` configuration
```

    + [Mocks](#mocks)
    + [Where will `SecretProvider` reside?](#where-will--secretprovider--reside-)
  * [Consequences](#consequences)

## Status

**proposed**

## Context

This ADR defines the new `SecretProvider` abstraction that will be used by all EdgeX services, including Device Services. The Secret Provider is used by services to retrieve secrets from the Secret Store. The Secret Store, in secure mode, is currently Vault. In non-secure mode it is configuration in some form, i.e. `DatabaseInfo` configuration or `InsecureSecrets` configuration for Application Services.

### Existing Implementations

The Secret Provider abstraction defined in this ADR is based on the Secret Provider abstraction implementations in the Application Functions SDK (App SDK) for Application Services and the one in go-mod-bootstrap (Bootstrap) used by the Core, Support & Security services in edgex-go. Device Services do not currently use secure secrets. The App SDK implementation was initially based on the Bootstrap implementation.

The similarities and differences between these implementations are:

- Both wrap the `SecretClient` from go-mod-secrets

> - Both initialize the `SecretClient` based on the `SecretStore` configuration(s)

Author: No, they use that for general communicate with Vault to create the SecrectStores, etc. The SecretProvider is all about  the services specific SecretStore.

- Both have factory method, but they differ greatly
- Both implement the `GetDatabaseCredentials` API
- Bootstrap's uses split interfaces definitions (`CredentialsProvider` & `CertificateProvider`) while the App SDK's use a single interface (`SecretProvider`) for the abstraction 

> - Bootstrap's includes the bootstrap handler while the App SDK's has the bootstrap handler separated out

Author: It is the one that the edgex-go services use.

- Bootstrap's implements the `GetCertificateKeyPair` API, which the App SDK's does not
- App SDK's implements the following, which the Bootstrap's does not
  - `Initialize` API (Bootstrap's initialization is done by the bootstrap handler)

>   - `StoreSecrets` API 

Author: Yes, this is how unknown secrets get seeded. Thing like Auth token for HTTP Export. Core services are known secrets that are specified in static configuration. Right now that is only DB Credentials.

**Service Shared** secrets are those that all instances of a class of service ,such a Application Services, share. Think of Core Data as it own class of service. An example of shared secrets are the database credentials for the single database instance for Store and Forward data that all Application Services may need to access. Another example is the database credentials for each of instance the Core Data. It is shared, but only one instance of Core Data is currently ever run. Service Shared secrets are seeded by security-secretstore-setup using static configuration for known secrets for known services. Currently database credentials are the only shared secrets. In the future we may have Message Bus credentials as shared secrets, but these will be truly shared secrets for all services to securely connect to the Message Bus, not just shared between instances of a service.

> 

Author: Yes.

#### Known and Unknown Services 

- **Known Services** are those identified in the static configuration by security-secretstore-setup
  
- These currently are Core Data, Core Metadata, Support Notifications, Support Scheduler and Application Service (class)
  
- **Unknown Services** are those not known in the static configuration that become known when added to the Docker compose file or Snap. 

  - Application Service (instance) are examples of these services. 

  - Service exclusive `SecretStore` can be created for these services by adding the services' unique name , i.e. appservice-http-export, to the `ADD_SECRETSTORE_TOKENS` environment variable for security-secretstore-setup

    ```
    ADD_SECRETSTORE_TOKENS: "appservice-http-export, appservice-mqtt-export"
    ```

    This creates an exclusive `SecretStore` for each service listed. The name provided for each service must be used in the service's `SecretStore` configuration and Docker volume mount. Typically the configuration is set via environment overrides or is already in an existing configuration profile (***http-export*** profile for app-service-configurable). 

> 

Author: In this case it is required. Will not work w/o the volume mount and the name has to match

    Example docker-compose file entries:

    ```yaml
    environment:
        ...
    	SecretStoreExclusive_Path: "/v1/secret/edgex/appservice-http-export/"
    	TokenFile: "/tmp/edgex/secrets/appservice-http-export/secrets-token.json"
    	
    volumes:
    	...
    	- /tmp/edgex/secrets/appservice-http-export:/tmp/edgex/secrets/appservice-http-export:ro,z
    ```

#### Known and Unknown Secrets

> - **Known Secrets** are those identified by name in the static configuration whose values are randomly generated at seed time. These secrets are seeded on start-up of EdgeX.

Author: We have this credentials generator:
https://github.com/edgexfoundry/edgex-go/blob/master/internal/security/secretstore/credentialgenerator.go#L30

>   - Database credentials are currently the only secrets of this type

Author: Definition added. Included in next commit.

> - **Unknown Secrets** are those not known in the static configuration and that become known during run time. These secrets are seeded at run time via the Application Services `/api/vX/Secrets` endpoint

Author: These secrets come for then user. They are external to Edgex, like HTTP endpoint auth token or external cloud MQTT broker credentials.

  -  HTTP header authorization credentials for HTTP Export are types of these secrets

#### Interfaces and factory methods

##### Bootstrap's current implementation

###### Interfaces

```go
type CredentialsProvider interface {
	GetDatabaseCredentials(database config.Database) (config.Credentials, error)
}
```

and

```go
type CertificateProvider interface {
	GetCertificateKeyPair(path string) (config.CertKeyPair, error)
}
```

###### Factory and bootstrap handler methods

```go
type SecretProvider struct {
	secretClient pkg.SecretClient
}

func NewSecret() *SecretProvider {
	return &SecretProvider{}
}

func (s *SecretProvider) BootstrapHandler(
	ctx context.Context,
	_ *sync.WaitGroup,
	startupTimer startup.Timer,
	dic *di.Container) bool {
    ...
    Intializes the SecretClient and adds it to the DIC for both interfaces.
    ...
}
```

##### App SDK's current implementation

###### Interface

```go
type SecretProvider interface {

> 	Initialize(_ context.Context) bool

Author: Nope. New Interface will have them.

```yaml
ADD_SECRETSTORE: "app-service-xyz[appservice/redisdb]"
```

> *Note: The secret identifier here is the short path to the secret in the existing **appservice**  `SecretStore`. In the above example this expands to the full path of `/secret/edgex/appservice/redisdb`*

The above example results in the Redis credentials being copied into app-service-xyz's `SecretStore` at `/secret/edgex/app-service-xyz/redis`.

Similar approach could be taken for Message Bus credentials where a common `SecretStore` is created with the Message Bus credentials saved. The services request the credentials are copied into their exclusive `SecretStore` using `common/messagebus` as the secret identifier.

Full specification for the environment variable's value is a comma separated list of service entries defined as:

```
<service-name>[optional list of known secret IDs sperated by ;]>
```

Example with one service specifying IDs for know secrets and one without known secrets 

```yaml

> ADD_SECRETSTORE: "appservice-xyz[appservice/redisdb; common/messagebus], appservice-http-export"

Reviewer 3: It is not clear to me where the credentials of common/messagebus are copied to?
I assume it will be /secret/edgex/app-service-xyz/messagebus
Should the first namespace be always replaced with the service name?

Reviewer: Let's make it more complex.

```
appservice-xyz[appservice/redisdb; common/brokers/messagebus], appservice-http-export
```

In this example, "messagebus" is the name of the secret, and "common/brokers" is the path to that secret relative to the root of the KV secret store.  This would create the secret "appservice-xyz/messagebox".  In other words, the last path component points at a secret that is going to be copied from one path to the base secrets dir for appservice-xyz.

Author: > It is not clear to me where the credentials of common/messagebus are copied to?
> I assume it will be /secret/edgex/app-service-xyz/messagebus
> Should the first namespace be always replaced with the service name?
Yes it is copied to `/secret/edgex/app-service-xyz/messagebus` and yes the `first namespace` is the service name.

Author: > ```
> appservice-xyz[appservice/redisdb; common/brokers/messagebus], appservice-http-export
> ```
Yes, we can make the path to the shared secret as complex as we feel is needed. All that matters is we know the path that is used to create the secret so we can specify it as in the above example.

Author: Yes, `security-secretstore-setup` would be responsible to seed any known shared secrets like it does today with DB credentials.

```

> When the `ADD_SECRETSTORE` environment variable is processed to create these `SecretStores`, it will copy the specified saved secrets from the initial `SecretStore` into the service's `SecretStore`. This all depends on the completion of database or other credential bootstrapping and the secrets having been stored prior to the environment variable being processed. security-secretstore-setup will need to be refactored to ensure this sequencing.

Reviewer 2: Q:  When to ensure this sequencing, will we introduce the circular dependency in startup?  Since the `security-secretstore-setup` need to wait for the `redis's dbcrendtials` secretStore to be seeded initially but at the same time `redis's` db container is also waiting for the `security-secretstore-setup` to signal done to proceed, right?  It looks like there is a race condition between these two containers: a catch-22 scenario to me OR at least this is not crystal clear to me on how this can happen today yet.

Reviewer: A.  Today, the code that processes ADD_SECRETSTORE runs before credential bootstrapping.  This needs to be changed for this to work.

A: Security secretstore setup today writes a flag SECRETSTORE_SETUP_DONE_FLAG that is watched by consul to signal that all of the secretstore bootstrapping is done.  It doesn't appear that redis-bootstrap blocks on this, or I am not seeing it.  Potential bug.

A: SECRETSTORE_SETUP_DONE_FLAG is going to be incorporated into the ready_to_run socket in the bootstrapping ADR so should be no concern there.

A: Potential issues with redis not remembering its provisioned password.

### Abstraction Interface

The following will be the new `SecretProvider` abstraction interface used by all Edgex services

```go
type SecretProvider interface {
    // Stores new secrets into the service's exclusive SecretStore at the speficied path.
	StoreSecrets(path string, secrets map[string]string) error

> 	// Retrieves secrets from the service's exclusive SecretStore at the speficied path.

Reviewer 2: ```suggestion
    // Retrieves secrets from the service's exclusive SecretStore at the speficied path.
```
align the code comment to look better

	GetSecrets(path string, _ ...string) (map[string]string, error)
    // Sets the secrets lastupdated time to current time. 
    // Needed for InsecureSecrets support
	SecretsUpdated()
    // Returns the secrets last updated time
	SecretsLastUpdated() time.Time
}
```

> *Note: The `GetDatabaseCredentials` and `GetCertificateKeyPair` APIs have been removed. These are no longer needed since insecure database credentials will no longer be stored in the `DatabaseInfo` configuration and certificate key pairs are secrets like any others. This allows these secrets to be retrieved via the `GetSecrets` API.*

### Implementation

#### Factory Method and Bootstrap Handler

The factory method and bootstrap handler will follow that currently in the Bootstrap implementation with some tweaks. Rather than putting the two split interfaces into the DIC, it will put just the single interface instance into the DIC. See details in the [Interfaces and factory methods](#interfaces-and-factory-methods) section above under **Existing Implementations**.

#### Caching of Secrets

> Secrets will be cached as they are currently in the Application Service implementation

Reviewer 4: How does the caching work? Is it the responsibility of the service to update the cache, or will it be handled automatically?

Author: Handled Automatically by the SecretProvider.

#### Insecure Secrets

Insecure Secrets will be handled as they are currently in the Application Service implementation. `DatabaseInfo` configuration will no longer be an option for storing the insecure database credentials. They will be stored in the `InsecureSecrets` configuration only.

```toml
[Writable.InsecureSecrets]
    [Writable.InsecureSecrets.DB]
    	path = "redisdb"
    	[Writable.InsecureSecrets.DB.Secrets]
            username = ""
            password = ""
```

##### Handling on-the-fly changes to `InsecureSecrets`

All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap` along with existing log level processing. This special processing will be taken from App SDK.

#### Mocks

> Proper mock of the `SecretProvider` interface will be created with `Mockery` to be used in unit tests. Current mock in App SDK is handle written rather then generated with `Mockery`.

Reviewer 2: ```suggestion
Proper mock of the `SecretProvider` interface will be created with `Mockery` to be used in unit tests. Current mock in App SDK is hand written rather then generated with `Mockery`.
```
hand written?

#### Where will `SecretProvider` reside?

##### Go Services

The final decision to make is where will this new `SecretProvider` abstraction reside? Originally is was assumed that it would reside in `go-mod-secrets`, which seems logical. If we were to attempt this with the implementation including the bootstrap handler, `go-mod-secrets` would have a dependency on `go-mod-bootstrap` which will likely create a circular dependency. 

Refactoring the existing implementation in `go-mod-bootstrap` and have it reside there now seems to be the best choice.

##### C Device Service

The C Device SDK will implement the same `SecretProvider` abstraction, InsecureSercets configuration and the underling `SecretStore` client.

### Consequences

> 

Reviewer 3: We might add one more consequence of the `go-mod-secrets` position.  Do we still need it?

Author: go-mod-secrets is still needed for the SecretClient that wraps the Vault requests.

- All service's will have `Writable.InsecureSecrets` section added to their configuration
- `InsecureSecrets` definition will be moved from App SDK to go-mod-bootstrap
- C SDK implementation could be big lift?
- edgex-go services will be modified to use the single `SecretProvider` interface from the DIC in place of current usage of the `GetDatabaseCredentials` and `GetCertificateKeyPair` interfaces.
  - Calls to `GetDatabaseCredentials` and `GetCertificateKeyPair` will be replaced with calls to `GetSecrets` API and appropriate processing of the returned secrets will be added. 
- App SDK will be modified to use `GetSecrets` API in place of the `GetDatabaseCredentials` API
- App SDK will be modified to use the new `SecretProvider` bootstrap handler

> - app-service-configurable's configuration profiles will be updated to remove the `SecretStoreExclusive` configuration and just use the existing `SecretStore` configuration

Reviewer 2: in addition, all the app-service-examples that are using this to-be-removed  `SecretStoreExclusive` need to be updated to accommodate the new configuration as well.

Author: Good ctach!

> - app-service-configurable's configuration profiles will be updated to remove the `SecretStoreExclusive` configuration and just use the existing `SecretStore` configuration

Reviewer 4: This will be an interesting case to handle in our configuration migration scripts for Ireland. I wonder if we need to start thinking about an ADR, issue, or some other means to start tracking breaking configuration changes being proposed?

Author: I have an action item for vF2F to start tracking this items... Haven't decided if use Wiki or Issue.

- security-secretstore-setup will be enhanced as described in the [Exclusive Secret Stores only](#exclusive-secret-stores-only) section above

> - Adding new services that need known secrets added to their `SecretStore` requires stopping and restarting all the services. The is because security-secretstore-setup will no longer know the values for the know secrets after it has completed setup.

Reviewer 4: Isn't the issue the security-secretstore-setup is no longer running (or waiting on /dev/null in docker) in order to handle the copy?

Author: Correct. It has finished and needs to rerun which then hoses the tokens and credentials for the other services.

      - [C Device Service](#c-device-service)
  * [Consequences](#consequences)

## Status

**proposed**

## Context

This ADR defines the new `SecretProvider` abstraction that will be used by all EdgeX services, including Device Services. The Secret Provider is used by services to retrieve secrets from the Secret Store. The Secret Store, in secure mode, is currently Vault. In non-secure mode it is configuration in some form, i.e. `DatabaseInfo` configuration or `InsecureSecrets` configuration for Application Services.

### Existing Implementations

The Secret Provider abstraction defined in this ADR is based on the Secret Provider abstraction implementations in the Application Functions SDK (App SDK) for Application Services and the one in go-mod-bootstrap (Bootstrap) used by the Core, Support & Security services in edgex-go. Device Services do not currently use secure secrets. The App SDK implementation was initially based on the Bootstrap implementation.

The similarities and differences between these implementations are:

- Both wrap the `SecretClient` from go-mod-secrets
- Both initialize the `SecretClient` based on the `SecretStore` configuration(s)

> - Both have factory method, but they differ greatly

Reviewer 4: sp: method --> **methods**

- Bootstrap's uses split interfaces definitions (`CredentialsProvider` & `CertificateProvider`) while the App SDK's use a single interface (`SecretProvider`) for the abstraction 
- Bootstrap's includes the bootstrap handler while the App SDK's has the bootstrap handler separated out
- Bootstrap's implements the `GetCertificateKeyPair` API, which the App SDK's does not
- App SDK's implements the following, which the Bootstrap's does not
  - `Initialize` API (Bootstrap's initialization is done by the bootstrap handler)
  - `StoreSecrets` API 
  - `GetSecrets` API
  - `InsecureSecretsUpdated` API
  - `SecretsLastUpdated` API
  - Wraps a second `SecretClient` for the Application Service instance's exclusive secrets.
    - Used by the `StoreSecrets` & `GetSecrets` APIs
  - The standard `SecretClient` is considered the shared client for secrets that all Application Service instances share. It is only used by the `GetDatabaseCredentials` API
  - Configuration based secret store for non-secure mode called `InsecureSecrets`
  - Caching of secrets
    - Needed so that secrets used by pipeline functions do not cause call out to Vault for every Event processed

#### What is a Secret?

> A secret is a collection of key/value pairs store in a `SecretStore` at specified path. Redis database credentials are an example of a `Secret` which contains the `username` and `password` key/values stored at the `redis` path.

Author: This was added to clarify the a Secret is not just a single value, but can be a collection of values. I have tweaked the wording.

> A secret is a collection of key/value pairs stored in a `SecretStore` at specified path whose values are sensitive in nature. Redis database credentials are an example of a `Secret` which contains the `username` and `password` key/values stored at the `redis` path.

Reviewer: ```suggestion
A secret is a collection of key/value pairs stored in a `SecretStore` at specified path whose values are sensitive in nature. Redis database credentials are an example of a `Secret` which contains the `username` and `password` key/values stored at the `redisdb` path.
```

> A secret is a collection of key/value pairs stored in a `SecretStore` at specified path whose values are sensitive in nature. Redis database credentials are an example of a `Secret` which contains the `username` and `password` key/values stored at the `redisdb` path.

Reviewer 4: This is probably out-of-scope for this ADR, but one thing that's not super clear to me is the fact that we have per service credentials (username/password) for Redis, but we also have a single password for Redis which is read by security-bootstrap-redis, which in turn sets this as the Redis password. I'm not quite sure how these credentials relate?

Author: Yes, out of scope for this ADR, but that is a Redis thing in that it didn't support the creation of a username as other DBs. I saw a not that Redis 6 does. So we should look at using that feature moving forward.

Reviewer: > Yes, out of scope for this ADR, but that is a Redis thing in that it didn't support the creation of a username as other DBs. I saw a note that Redis 6 does. So we should look at using that feature moving forward.

I read the [docs](https://redis.io/topics/acl) and I am not convinced the new ACL system supports persistent usernames and passwords.  It still seems to be dynamic or relying on either redis.conf or an external ACL file.

#### Service Exclusive vs Service Shared Secrets

> **Service Exclusive** secrets are those that are exclusive to the instance of the running service. An example of exclusive secrets are the HTTP Auth tokens used by two running instances of app-service-configurable (http-export) which export different device Events to different endpoints with different Auth tokens in the HTTP headers.  Service Exclusive secrets are seeded by POSTing the secrets to the `/api/vX/Secrets` endpoint on the running instance of each Application Service.

Reviewer: ```suggestion
**Service Exclusive** secrets are those that are exclusive to the instance of the running service. An example of exclusive secrets are the HTTP Auth tokens used by two running instances of app-service-configurable (http-export) which export different device Events to different endpoints with different Auth tokens in the HTTP headers.  Service Exclusive secrets are seeded by POSTing the secrets to the `/api/vX/secrets` endpoint on the running instance of each Application Service.
```

> **Service Exclusive** secrets are those that are exclusive to the instance of the running service. An example of exclusive secrets are the HTTP Auth tokens used by two running instances of app-service-configurable (http-export) which export different device Events to different endpoints with different Auth tokens in the HTTP headers.  Service Exclusive secrets are seeded by POSTing the secrets to the `/api/vX/Secrets` endpoint on the running instance of each Application Service.

Reviewer: If one ran two instances of http-export, would that count as a service shared secret or a service exclusive secret?

Author: Exclusive as the only reason to run 2 is to have them export to different endpoint and thus need different secrets.

> **Service Exclusive** secrets are those that are exclusive to the instance of the running service. An example of exclusive secrets are the HTTP Auth tokens used by two running instances of app-service-configurable (http-export) which export different device Events to different endpoints with different Auth tokens in the HTTP headers.  Service Exclusive secrets are seeded by POSTing the secrets to the `/api/vX/secrets` endpoint on the running instance of each Application Service.

Reviewer 4: Is there anyway for an application service to autogenerate a secret itself at runtime and provision it vs. requiring an additional service/program to POST the secret? I suppose the service could just POST to itself?

Author: No, these types of secrets are generated by the 3rd party that the service is exporting to. If an AWS access token, etc. Those can't not be autogenerated by the POSTing client. ;-)

> **Service Shared** secrets are those that all instances of a class of service ,such a Application Services, share. Think of Core Data as it own class of service. An example of shared secrets are the database credentials for the single database instance for Store and Forward data that all Application Services may need to access. Another example is the database credentials for each of instance the Core Data. It is shared, but only one instance of Core Data is currently ever run. Service Shared secrets are seeded by security-secretstore-setup using static configuration for known secrets for known services. Currently database credentials are the only shared secrets. In the future we may have Message Bus credentials as shared secrets, but these will be truly shared secrets for all services to securely connect to the Message Bus, not just shared between instances of a service.

Author: Not finding matches for these. More context?

Application Services currently have the ability to configure `SecretStores` for **Service Exclusive** and/or **Service Shared** secrets depending on their needs.

#### Known and Unknown Services 

- **Known Services** are those identified in the static configuration by security-secretstore-setup
  
  - These currently are Core Data, Core Metadata, Support Notifications, Support Scheduler and Application Service (class)
  
- **Unknown Services** are those not known in the static configuration that become known when added to the Docker compose file or Snap. 

  - Application Service (instance) are examples of these services. 

>   - Service exclusive `SecretStore` can be created for these services by adding the services' unique name , i.e. appservice-http-export, to the `ADD_SECRETSTORE_TOKENS` environment variable for security-secretstore-setup

Reviewer: Should specify which service needs the env var.

Author: Already do. `environment variable for security-secretstore-setup`

    ```
    ADD_SECRETSTORE_TOKENS: "appservice-http-export, appservice-mqtt-export"
    ```

>     This creates an exclusive `SecretStore` for each service listed. The name provided for each service must be used in the service's `SecretStore` configuration and Docker volume mount. Typically the configuration is set via environment overrides or is already in an existing configuration profile (***http-export*** profile for app-service-configurable). 

Reviewer: ```suggestion
    This creates an exclusive secret store token for each service listed. The name provided for each service must be used in the service's `SecretStore` configuration and Docker volume mount. Typically the configuration is set via environment overrides or is already in an existing configuration profile (***http-export*** profile for app-service-configurable). 
```

>     This creates an exclusive secret store token for each service listed. The name provided for each service must be used in the service's `SecretStore` configuration and Docker volume mount. Typically the configuration is set via environment overrides or is already in an existing configuration profile (***http-export*** profile for app-service-configurable). 

Reviewer 4: nit: Docker volume mount (**if applicable**)

    Example docker-compose file entries:

    ```yaml
    environment:
        ...
    	SecretStoreExclusive_Path: "/v1/secret/edgex/appservice-http-export/"
    	TokenFile: "/tmp/edgex/secrets/appservice-http-export/secrets-token.json"
    	
    volumes:
    	...
    	- /tmp/edgex/secrets/appservice-http-export:/tmp/edgex/secrets/appservice-http-export:ro,z
    ```

#### Known and Unknown Secrets

- **Known Secrets** are those identified by name in the static configuration whose values are randomly generated at seed time. These secrets are seeded on start-up of EdgeX.
  - Database credentials are currently the only secrets of this type

> - **Unknown Secrets** are those not known in the static configuration and that become known during run time. These secrets are seeded at run time via the Application Services `/api/vX/Secrets` endpoint

Reviewer 5: Known and Unknown names here seem confusing to me.  What about SeedTimeSecrets and RuntimeSecrets?  Both are randomly generated - it is just a matter of when.

Author: Nope, they are not both randomly generated, Unknow Secrets are things like and Auth token for REST call to export data to an endpoint. Both are seeded, static Known ones at startup and Unknown ones when ever the user pushes (seeds) the secret via the `Secrets` endpoint.

How about `Static Defined Secrets` and `Runtime Secrets`?

Author: Changed to  `Static Secrets and Runtime Secrets`

> - **Runtime Secrets** are those not known in the static configuration and that become known during run time. These secrets are seeded at run time via the Application Services `/api/vX/Secrets` endpoint

Reviewer: ```suggestion
- **Runtime Secrets** are those not known in the static configuration and that become known during run time. These secrets are seeded at run time via the Application Services `/api/vX/secrets` endpoint
```

```yaml
ADD_SECRETSTORE: "app-service-xyz[appservice/redisdb]"
```

> *Note: The secret identifier here is the short path to the secret in the existing **appservice**  `SecretStore`. In the above example this expands to the full path of `/secret/edgex/appservice/redisdb`*

The above example results in the Redis credentials being copied into app-service-xyz's `SecretStore` at `/secret/edgex/app-service-xyz/redis`.

Similar approach could be taken for Message Bus credentials where a common `SecretStore` is created with the Message Bus credentials saved. The services request the credentials are copied into their exclusive `SecretStore` using `common/messagebus` as the secret identifier.

Full specification for the environment variable's value is a comma separated list of service entries defined as:

```

> <service-name>[optional list of static secret IDs sperated by ;]>

Reviewer 2: - syntax-wise, there is an extra `>` at the end of this example
- should have another `service-name` in this example to fit the aforementioned sentence for `comma separated list`

```suggestion
<service-name1>[optional list of static secret IDs sperated by ;],<service-name2>[optional list of static secret IDs sperated by ;],...
```

Author: I Like it!

- `InsecureSecrets` definition will be moved from App SDK to go-mod-bootstrap
- Go Device SDK will add the SecretProvider to it's bootstrapping 
- C Device SDK implementation could be big lift?
- ` SecretStore`configuration section will be added to all Device Services
- edgex-go services will be modified to use the single `SecretProvider` interface from the DIC in place of current usage of the `GetDatabaseCredentials` and `GetCertificateKeyPair` interfaces.
  - Calls to `GetDatabaseCredentials` and `GetCertificateKeyPair` will be replaced with calls to `GetSecrets` API and appropriate processing of the returned secrets will be added. 
- App SDK will be modified to use `GetSecrets` API in place of the `GetDatabaseCredentials` API
- App SDK will be modified to use the new `SecretProvider` bootstrap handler
- app-service-configurable's configuration profiles as well as all the Application Service examples configurations will be updated to remove the `SecretStoreExclusive` configuration and just use the existing `SecretStore` configuration
- security-secretstore-setup will be enhanced as described in the [Exclusive Secret Stores only](#exclusive-secret-stores-only) section above

> - Adding new services that need static secrets added to their `SecretStore` requires stopping and restarting all the services. The is because security-secretstore-setup has completed but not stopped. If it is rerun without stopping the other services, there tokens and static secrets will have changed. The planned refactor of `security-secretstore-setup` will attempt to resolved this.

Reviewer: ```suggestion
- Adding new services that need static secrets added to their `SecretStore` requires stopping and restarting all the services. The is because security-secretstore-setup has completed but not stopped. If it is rerun without stopping the other services, there tokens and static secrets will have changed. The planned refactor of `security-secretstore-setup` will attempt to resolve this.
```

Author: type fixed locally so when updated to `Approved` this will be included.

> - Snaps does not yet support setting the environment variable for adding SecretStore. It is planned Ireland.

Reviewer 2: ```suggestion
- Snaps does not yet support setting the environment variable for adding SecretStore. It is planned for Ireland release.
```

> - Snaps does not yet support setting the environment variable for adding SecretStore. It is planned for Ireland release.

Reviewer 4: nit: Snaps **do** not yet...
